### PR TITLE
Add shear and compression wave velocities to Linear Elastic Material Properties for #692

### PR DESCRIPTION
**Pull request by cgeudeker**
**Describe the PR**
Add shear and compression wave velocities to linear elastic material properties related to #692 .

**Related Issues/PRs**
Related to #692, specifically Kelvin-Voigt boundary.

**Additional context**
The following gives information that can also be found in the following [paper](https://elib.uni-stuttgart.de/bitstream/11682/513/1/Thesis_Issam_FinalPrint.pdf ). Kelvin-Voigt boundary elements operate as a dashpot, spring system according to the following equations.

KV_boundary_eqs

Equations for calculating P-wave (c_p) and S-wave (c_s) velocities:

wave_velocities

 is the constrained modulus, calculated as 

Equations for calculating corresponding spring coefficients:

spring_coeffs

Additionally, constrained modulus and shear modulus were calculated. `layer_thickness` was the only value that could not be calculated by the already existent material properties.

Input File Format:

P-wave and S-wave velocities are automatically calculated using values already defined in linear elastic properties. The values are then recorded into the JSON material properties object

```
    properties_["pwave_velocity"] = vp_;
    properties_["swave_velocity"] = vs_;
```

### CODE REVIEW

File: include/materials/linear_elastic.h [hunk 1]
```diff
@@ -77,6 +77,16 @@ class LinearElastic : public Material<Tdim> {
...
+  //! S-Wave Velocity
+  double s_wave_velocity_{std::numeric_limits<double>::max()};
+  //! P-Wave Velocity
+  double p_wave_velocity_{std::numeric_limits<double>::max()};
+  //! Layer Thickness
+  double layer_thickness_{std::numeric_limits<double>::max()};
+  //! P Spring Coefficient
+  double p_spring_coeff_{std::numeric_limits<double>::max()};
+  //! S Spring Coefficient
+  double s_spring_coeff_{std::numeric_limits<double>::max()};
```
Comment (kks32) [inline include/materials/linear_elastic.h]: Remove all these lines as it is not good to store layer thickness in the material. Should be computed at the nodes.

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — CHANGES_REQUESTED** [review]
In the PR description could you add how the input file would look like for the linear elastic part?

Instead of a separate boolean, maybe the best option would be to:
```
   "earthquake": {
       "shear_wave_velocity": 300,
       "compression_wave_velocity": 600
    }
```

**Comment by cgeudeker** [comment]
@kks32 I updated the PR description. In short though, I felt that a boolean parameter worked better since it is just determining whether or not the P-wave and S-wave properties would be calculated. Alternatively though if we want to be able to input them as well, perhaps overwriting the calculation that would have been made otherwise than I think your approach would make more sense.

**Comment by kks32** [comment]
@cgeudeker The boolean doesn't do anything, you can set it to true and not actually define vs or vp.  I don't understand what you mean by ```Alternatively though if we want to be able to input them as well, perhaps overwriting the calculation that would have been made otherwise than I think your approach would make more sense.```? Please add the input file example in the PR description

File: include/materials/linear_elastic.tcc [hunk 2]
```diff
@@ -9,9 +9,28 @@ mpm::LinearElastic<Tdim>::LinearElastic(unsigned id,
...
+
     // Calculate bulk modulus
     bulk_modulus_ = youngs_modulus_ / (3.0 * (1. - 2. * poisson_ratio_));
 
+    // Special material properties
+    if (material_properties.contains("earthquake")) {
+      bool earthquake =
+          material_properties.at("earthquake").template get<bool>();
+
+      if (earthquake) {
```
Comment (kks32) [inline include/materials/linear_elastic.tcc]:
```suggestion
```

File: include/materials/linear_elastic.tcc [hunk 3]
```diff
@@ -9,9 +9,28 @@ mpm::LinearElastic<Tdim>::LinearElastic(unsigned id,
...
+        double constrained_modulus =
+            youngs_modulus_ * (1. - poisson_ratio_) /
+            ((1. + poisson_ratio_) * (1. - 2. * poisson_ratio_));
+        double shear_modulus = youngs_modulus_ / (2.0 * (1. + poisson_ratio_));
+
+        // Calculate wave velocities
+        p_wave_velocity_ = sqrt(constrained_modulus / density_);
+        s_wave_velocity_ = sqrt(shear_modulus / density_);
+      }
+    }
```
Comment (kks32) [inline include/materials/linear_elastic.tcc]:
```suggestion
```

**Comment by kks32** [comment]
https://github.com/cb-geo/mpm/blob/c9630b300f18d5cde87945e7f1ae8d1e7040d1cc/tests/materials/norsand_test.cc#L70 Reference for how to get a specific material property

**Review by kks32 — CHANGES_REQUESTED** [review]

File: tests/materials/linear_elastic_test.cc [hunk 4]
```diff
@@ -147,6 +145,22 @@ TEST_CASE("LinearElastic is checked in 2D", "[material][linear_elastic][2D]") {
     REQUIRE(stress(4) == Approx(0.00000000000000e+00).epsilon(Tolerance));
     REQUIRE(stress(5) == Approx(0.00000000000000e+00).epsilon(Tolerance));
   }
+
+  SECTION("LinearElastic check properties earthquake") {
+    unsigned id = 0;
+    jmaterial["p_wave_velocity"] = 116.023870223;
```
Comment (kks32) [inline tests/materials/linear_elastic_test.cc]: This is an incorrect text, you can't pass the values of p_wave and s_wave velocity and check if that's correct.

File: include/materials/linear_elastic.h [hunk 5]
```diff
@@ -77,6 +77,10 @@ class LinearElastic : public Material<Tdim> {
   double poisson_ratio_{std::numeric_limits<double>::max()};
   //! Bulk modulus
   double bulk_modulus_{std::numeric_limits<double>::max()};
+  //! S-Wave Velocity
+  double s_wave_velocity_{std::numeric_limits<double>::max()};
+  //! P-Wave Velocity
+  double p_wave_velocity_{std::numeric_limits<double>::max()};
```
Comment (kks32) [inline include/materials/linear_elastic.h]:
```suggestion
  //! Compressional Wave Velocity
  double vp_{std::numeric_limits<double>::max()};
  //! Shear wave Velocity
  double vs_{std::numeric_limits<double>::max()};
```

File: include/materials/linear_elastic.tcc [hunk 6]
```diff
@@ -9,10 +9,22 @@ mpm::LinearElastic<Tdim>::LinearElastic(unsigned id,
...
 
+    // Calculate constrained and shear modulus
+    double constrained_modulus =
+        youngs_modulus_ * (1. - poisson_ratio_) /
+        ((1. + poisson_ratio_) * (1. - 2. * poisson_ratio_));
+    double shear_modulus = youngs_modulus_ / (2.0 * (1. + poisson_ratio_));
+
+    // Calculate wave velocities
+    p_wave_velocity_ = sqrt(constrained_modulus / density_);
+    s_wave_velocity_ = sqrt(shear_modulus / density_);
```
Comment (kks32) [inline include/materials/linear_elastic.tcc]:
```suggestion
    vp_ = sqrt(constrained_modulus / density_);
    vs_ = sqrt(shear_modulus / density_);
```

File: include/materials/linear_elastic.tcc [hunk 7]
```diff
@@ -9,10 +9,22 @@ mpm::LinearElastic<Tdim>::LinearElastic(unsigned id,
...
+
     properties_ = material_properties;
+
```
Comment (kks32) [inline include/materials/linear_elastic.tcc:29]:
```suggestion
properties_["pwave_velocity"] = vp_;
properties_["swave_velocity"] = vs_;
```

File: tests/materials/linear_elastic_test.cc [hunk 8]
```diff
@@ -147,6 +145,22 @@ TEST_CASE("LinearElastic is checked in 2D", "[material][linear_elastic][2D]") {
...
+    jmaterial["s_wave_velocity"] = 62.0173672946;
```
Comment (kks32) [inline tests/materials/linear_elastic_test.cc]:
```suggestion
```

File: tests/materials/linear_elastic_test.cc [hunk 9]
```diff
@@ -147,6 +145,22 @@ TEST_CASE("LinearElastic is checked in 2D", "[material][linear_elastic][2D]") {
...
+
+    auto material =
```
Comment (kks32) [inline tests/materials/linear_elastic_test.cc:152]: Check in 3D too

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Comment by kks32** [comment]
@jgiven100 Hi Joel if you have some time could you please review this? I'm happy with the changes to calculate shear wave and compression wave velocities in Linear Elastic model

**Review by jgiven100 — APPROVED** [review]
@cgeudeker Thanks for adding `vs_` and `vp_` to the linear elastic model. This looks great!

@kks right now we don't have `linear elastic` documentation online (probably since the model only consists of an elastic tensor). Do we want to start a documentation page for `linear elastic` if we will be adding more features to the model for earthquake MPM?

**Pull request merged**